Feature/2278 update solutions ui

ChengYanJin opened the pull request:
**Component**: ui, solution

**Context**:  Because we change a lot of stuff on how we manage Solutions in #2277

**Summary**:
 - From UI, I am able to create an environment
- From UI,  I am able to add the solution from the available solutions in the specific environment.
- If the solution deployment is done,  we should be able to see the solution name with version displayed in the specific environment row
- From UI, I am able to delete the Environment by click on the trash button 

**Acceptance criteria**: 

1. build the solution ISO

   Go to `bugfix/fix-example-solution` branch to make iso.

 - To get a new version of the same Solution, I edited the VERSION file
 - To get a "new" Solution, I passed PRODUCT_NAME="Other Solution" PRODUCT_LOWERNAME="other-solution" to make iso

   Copy the iso to `_build/root`

2. import the solution ISO
`./solutions.sh import --archive </path/to/solution.iso>`
3. activate the solution:
`./solutions.sh activate --name <solution-name> --version <solution-version>`
4. create an environment
5. click on the add a solution button

screenshots:
1. create a new environment: [image1](https://user-images.githubusercontent.com/18453133/77231089-f3573080-6b98-11ea-93ac-d3c633e0d652.png)
2. add a solution: [image2](https://user-images.githubusercontent.com/18453133/77231213-da9b4a80-6b99-11ea-8c79-8ede3ea73df4.png)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2278 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
ui/api: Split K8s APIs by domain

To help with organization of the increasing number of API call helpers,
we split them into domain-based modules, e.g. `k8s/solutions` or
`k8s/core`.

Refs: #2278

gdemonet committed:
ui/solutions: Remove UI services from state

We used to consider Admin UIs as unique per Solution across a cluster.
That's however not the case anymore, so we remove this data from the
Redux state.

Issue: #2278

gdemonet committed:
ui/solutions: Rework environments API

The environments are not exposed using CustomObjects anymore.
Instead, we manage them through labels on namespaces.

Refs: #2278

gdemonet committed:
ui/solutions: Read Solutions info in environments

We add more data about configured Solutions in an environment (through a
'metalk8s-environment' ConfigMap), and the Admin UI for each.
This will be used in the Solutions page to display Environments info.

This changes from the previous approach, as Admin UIs used to be unique
per Solution, not per Solution in each environment.

Refs: #2278

gdemonet committed:
ui/solutions: Simplify Solutions ConfigMap handling

The 'metalk8s-solutions' ConfigMap is unique in the cluster, so we can
simply `GET` it (instead of listing and filtering based on name).

Refs: #2278

gdemonet committed:
ui/solutions: Highlight active Solution versions

This makes the "active" version of a Solution stand out from the others.
An improvement would be to show a tooltip for this column, explaining
that an "active" version is the one for which cluster-wide components
(such as storage classes or CRDs) are deployed.

Refs: #2278

gdemonet committed:
ui/solutions: Rework listing of solutions and envs

We update the display of Solutions and Environments tables based on the
new data available in the Redux state.

Refs: #2278

gdemonet committed:
ui/solutions: Rework refresh saga

Use a more explicit pattern for the refresh saga, based on `take`,
`fork` and `race`.

See: #2035

gdemonet committed:
ui/solutions: Refresh Solutions only when on Solutions page

ChengYanJin committed:
ui/solutions: Fix the intl error of solutionlist

- since we have import the intl from the IntlGlobalProvider
  we don't need to get the intl from props

Refs: #2278

ChengYanJin committed:
ui/translation: Update the translations

- rename from solutions-> solution versions->version for add a solution modal

Refs: #2278

ChengYanJin committed:
ui/k8sapi: Add some k8s api for the configMap and Namespace

Refs: #2278

ChengYanJin committed:
ui/solutions: Update the solution saga

Refs: #2278

ChengYanJin committed:
ui/solutions: Add name prop for salt job

- since we want to track the job of the salt, we could add prefix e.g.'prepare-env/'
  to the salt job when it starts, and remove the prefix when is done.

Refs: #2278

ChengYanJin committed:
ui/solutions: Add deployed solution in the evironment table

- add the loader when the salt starts it's job

Refs: #2278

ChengYanJin committed:
ui/package: Fix the warning of eslint by update the version

- ESLint 6.2.0 + babel-eslint + no-unused-vars false positive with for-in loop
  https://github.com/babel/babel-eslint/issues/791

Refs: #2278

ChengYanJin committed:
ui/package-lock: Update the package-lock

Refs: #2278

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

ChengYanJin committed:
ui/test: Update the saga test for saga

- since we add name for the job so that we could know the status of the job by name

Refs: #2278

ChengYanJin committed:
ui/test: Add unit test for solution

Refs: #2278

gdemonet reviewed (commented):
Some general comments first, related to previous commits you had to work on top of:

- there are still some commits left as `WIP`: probably worth to rename them if they are now valid (from your standpoint), or simply remove them if irrelevant
- the `[WIP] ui: Run app behind /metalk8s` shouldn't be done in this PR I believe, not related to the Solutions page IMHO (would make sense in some "Shell integration" PR though :wink:) (the `serve fonts CSS...` commit should move out as well)

gdemonet commented on ui/src/translations/en.json:
```diff
@@ -135,5 +136,10 @@
+  "documentation": "Documentation",
+  "env_preparation": "Environment preparation",
+  "env_preparation_success": "Environment {envName} is being prepared.",
+  "env_preparation_failed": "Environemnt {envName} preparation failed.",
```
```suggestion
  "env_preparation_failed": "Environment {envName} preparation failed.",
```

slaperche-scality replied:
`has failed` (to be consistent with the message for `environment_creation_failed`)

gdemonet commented on ui/src/translations/en.json:
```diff
@@ -135,5 +136,10 @@
+  "env_preparation": "Environment preparation",
+  "env_preparation_success": "Environment {envName} is being prepared.",
+  "env_preparation_failed": "Environemnt {envName} preparation failed.",
+  "preparing_environemnt": "preparing the environment",
```
```suggestion
  "preparing_environment": "preparing the environment",
```

gdemonet replied:
Also, not sure why we need this key (aren't `env_preparation` / `env_preparation_success` sufficient?)

ChengYanJin replied:
ok. At the beginning, I would like to have this prop used for the text of the loader.
But yeah, it makes sense to reuse`env_preparation_success`

gdemonet replied:
Oh OK I see! So no, let's keep a loading message as this one.

However, `env_preparation_success` should show a "successful" status, no? If yes, maybe reword to show:
- in English: `Environment {envName} preparation succeeded.`
- in French: `La préparation de l'environnement {envName} a réussi.`

The current message for `_success` is actually meaning something is `_in_progress` ;)

ChengYanJin replied:
Ok. For the loading message, I will keep the previous one.

However, I don't think it's a good idea to change the text of `env_preparation_success`. Probably we should change the key to `env_preparation_in_progress` though.

I am going to give the context. The text will be displayed as the notification after the Salt job is done. So in theory, we couldn't say it's already succeeded because we still need to check the deployment of the solution.

So IMO, `Environment {envName} is being prepared.` is better. What do you think?

gdemonet replied:
> The text will be displayed as the notification after the Salt job is done. So in theory, we couldn't say it's already succeeded because we still need to check the deployment of the solution.

That's where I disagree ;) When Salt returns after the `prepare-environment`, it means (at least, from Salt PoV) that the env has been prepared.
It may not mean that all resources (UI/Operator) are available in the current implem, but I'd rather add this check on Salt side than needing an extra step in UI or CLI.

gdemonet commented on ui/src/translations/en.json:
```diff
@@ -135,5 +136,10 @@
+  "env_preparation_success": "Environment {envName} is being prepared.",
+  "env_preparation_failed": "Environemnt {envName} preparation failed.",
+  "preparing_environemnt": "preparing the environment",
+  "environment_creation_failed": "Environment {name} creation has failed."
```
For consistency with other keys:

```suggestion
  "env_creation_failed": "Environment {envName} creation has failed."
```

gdemonet commented on ui/src/translations/fr.json:
```diff
@@ -135,5 +136,10 @@
   "redirecting": "Redirection…",
-  "documentation": "Documentation"
+  "documentation": "Documentation",
+  "env_preparation": "Environment preparation",
```
```suggestion
  "env_preparation": "Préparation d'environnement",
```

slaperche-scality replied:
Probably not a correct French translation, what's the context/use for this string?
`Préparation de l’environnement` maybe

gdemonet commented on ui/src/translations/fr.json:
```diff
@@ -135,5 +136,10 @@
+  "env_preparation": "Environment preparation",
+  "env_preparation_success": "L'environnement {envName} est en cours de préparation.",
+  "env_preparation_failed": "La préparation de l'environnement {envName} a échoué.",
+  "preparing_environemnt": "préparer l'environnement",
```
If this is supposed to mean the same thing as the english variant, it should translate to:

```suggestion
  "preparing_environemnt": "environnement en cours de préparation",
```

slaperche-scality commented on ui/src/translations/fr.json:
```diff
@@ -123,7 +124,7 @@
   "description": "Description",
   "add_solution_to_environment": "Ajout d'une solution à l'environnement {environment}",
-  "version_env": "Versions",
+  "version_env": "Version",
```
Why `version_env` instead of `version`?

ChengYanJin replied:
because we have another translation  `"version": "MetalK8s Version",`

slaperche-scality replied:
Then `"version": "MetalK8s Version"` should probably be renamed `"metalk8s_version": "MetalK8s Version"` instead

slaperche-scality commented on ui/src/translations/fr.json:
```diff
@@ -135,5 +136,10 @@
+  "env_preparation": "Environment preparation",
+  "env_preparation_success": "L'environnement {envName} est en cours de préparation.",
+  "env_preparation_failed": "La préparation de l'environnement {envName} a échoué.",
+  "preparing_environemnt": "préparer l'environnement",
```
This one is a bit weird too, what's the use/context?

If it's for a progress bar or some kind of load, then `Préparation de l’environnement…` would be better (and English could be `Preparing the environment…`)

ChengYanJin replied:
Thanks for the French speaker^^
The context of `preparing_environemnt` is that we want to display a text next to the loader when we are waiting for preparing the environment.

gdemonet commented on ui/src/services/k8s/solutions.js:
```diff
@@ -96,18 +98,57 @@ export async function getEnvironmentAdminUIs(environment) {
-    if (!result.error) {
-      // Return the first one found among the Environment namespaces
-      return { ...result?.data };
-    }
```
Not sure this is a good change. Under `services/k8s/solutions`, the goal is to have methods that are specific to Solutions, while keeping generic things under `services/k8s/core` (but could also be split up and extended if required).

This method is retrieving the "environment ConfigMap" for a given environment. It's built on the premise that a caller has a access to an "environment" object, as returned by `listEnvironments` above.
Such "environments" can have multiple `namespaces`, and this is why the `getEnvironmentConfigMap` goes through this list and returns the first CM it finds. Not the best solution (merging them into a higher-level structure would be great, but takes time, so, later :innocent:), but at least we have room for improvement directly in this helper method.

If you really want helpers around `get/patch/create` for ConfigMaps, please add them to `services/k8s/core` instead, and build on them in this file.

ChengYanJin replied:
Good point!!! We may have multiple namespaces in one environment. I will just as you said leave space for future improvement, anyway in UI we don't support multiple namespaces yet.

slaperche-scality commented on ui/src/translations/en.json:
```diff
@@ -96,6 +96,7 @@
   "no_storage_class_found": "No storage class found.",
   "storage_class_is_required": "A storage class is required in order to create a volume. Please create one in Kubernetes.",
   "learn_more": "Learn more",
+  "solution": "Solution",
```
Don't we have something to handle the pluralization in the framework we use fo the translation?

Edit: Yeah, not sure that would help in that case.

gdemonet commented on ui/src/services/k8s/solutions.js:
```diff
@@ -96,18 +98,57 @@ export async function getEnvironmentAdminUIs(environment) {
+  } catch (error) {
+    return { error };
   }
 }
```
Creating an environment ConfigMap could benefit from some pre-filled `data` keys in the future, so let's make this a helper method:

```javascript
export async function createEnvironmentConfigMap(namespace) {
  // Add default values in here
  const data = {};

  // Note to YanJin: this `core.createConfigMap` helper would need
  // to be defined in `services/k8s/core`
  return core.createConfigMap(ENVIRONMENT_CONFIGMAP_NAME, namespace, data);
}
```

gdemonet commented on ui/src/services/k8s/solutions.js:
```diff
@@ -96,18 +98,57 @@ export async function getEnvironmentAdminUIs(environment) {
+  } catch (error) {
+    return { error };
+  }
+}
```
Same principles here: define the generic "patch ConfigMap" method in `core`, and build on it in `solutions` for just your needs!

```javascript
export async function addSolutionToEnv(namespace, solName, solVersion) {
  const patch = { op: "add", path: `/data/${solName}`, value: solVersion };

  return core.patchConfigMap(
    ENVIRONMENT_CONFIGMAP_NAME, namespace,
    // use an object instead of many arguments, for easier support of
    // multiple methods (e.g. both `jsonPatch` and `mergePatch` could be
    // allowed, headers being inferred from which is used).
    { jsonPatch = [patch] },
  );
}

export async function updateSolutionInEnv(namespace, solName, solVersion) {
  const patch = { op: "replace", path: `/data/${solName}`, value: solVersion };

  return core.patchConfigMap(
    ENVIRONMENT_CONFIGMAP_NAME, namespace, { jsonPatch = [patch] },
  );
}

export async function removeSolutionFromEnv(namespace, solName) {
  const patch = { op: "remove", path: `/data/${solName}` };

  return core.patchConfigMap(
    ENVIRONMENT_CONFIGMAP_NAME, namespace, { jsonPatch = [patch] },
  );
}
```

ChengYanJin replied:
Good to know the jsonPatch!!!

gdemonet commented on ui/src/services/k8s/solutions.js:
```diff
@@ -96,18 +98,57 @@ export async function getEnvironmentAdminUIs(environment) {
+  } catch (error) {
+    return { error };
+  }
+}
```
Definitely a helper to put in `core` (or `app`, if you want to add one, but I'd say `core` is for everything "core" K8s, while others are for Metal-specific concepts).

gdemonet reviewed (changes requested):
Early dump of comments, mostly related to the `services/k8s` helpers.

IMO, we should make sure that:

- as much business logic for managing APIs (such as K8s or Prom) as possible should go into the `services` helpers
- ducks are UI logic: "when this action was dispatched, trigger this call to a `services` helper, and let it do the business logic" - "when this call completed, store the result in the state, and trigger a notification"
- we don't put some logic in both places

The special logic we need for Solutions and Volumes are worth extracting entirely into `services/k8s/solutions` and `s/k/volumes` respectively. Ducks can then rely on the helpers, making sure that if for some reason we change the backend implementation of any of them, the helpers can hide this change behind a constant format, and ducks will remain unchanged.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

ChengYanJin committed:
ui/saltAPI: Add prepareEnvironment for salt services

Refs: #2278

ChengYanJin committed:
ui/services: Rearrange the services helpers

- we should put solution logic in services/k8s/solution while
  generic logic for metalk8s in services/k8s/core

Refs: #2278

ChengYanJin committed:
ui/solutions: Update the text for the loader in solutionList

Refs: #2278

ChengYanJin committed:
ui/ducks: Extract the updateEnvironment logic

- Every time when we refresh the environment, we should call the updateEnvironment to
  check the Operator & UI deployment and update the solutions list in the Environemnt

Refs: #2278

ChengYanJin committed:
ui/translations: Update the key from version to metalk8s_version

- Since we have the version in Environment table with version key for translation
  it's better to rename the version to metalk8s_version

Refs: #2278

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -65,52 +78,184 @@ export function createEnvironmentAction(newEnvironment) {
+  const jobs = yield select(state => state.app.salt.jobs);
+  const preparingEnvs = jobs
+    ?.filter(job => job.name.startsWith('prepare-env/') && !job.completed)
+    .map(job => job.name.replace(/^(prepare-env\/)/, ''));
```
Can we use `job.type === "prepare-env"` instead of prefixing the `job.name`, as we did for Node deployments?

ChengYanJin committed:
ui/solutions: Delelte the Environment

- add the trash button in the solutionList
- call the coreV1.deleteEnvironment api

Refs: #2278

gdemonet commented on ui/package.json:
```diff
@@ -7,7 +7,9 @@
     "@kubernetes/client-node": "github:scality/kubernetes-client-javascript.git#browser-0.10.2-62-g61d8543",
     "@scality/core-ui": "github:scality/core-ui.git#v0.5.0",
     "axios": "^0.18.0",
+    "eslint": "6.8.0",
```
Is this really a project dependency? Or just one for `devDepencies`? ;)

gdemonet commented on ui/src/containers/SolutionList.js:
```diff
@@ -60,13 +66,14 @@ const FormStyle = styled.div`
+  /* solve the tooltip display issue */
+  .sc-table-column-cell-action {
+    overflow: visible !important;
+  }
```
Something wrong with this? Ticket in `core-ui` maybe? If there's one already, having it in the comment can help :)

ChengYanJin replied:
Yeah. I just create a ticket in core-ui: https://github.com/scality/core-ui/issues/176
Will put it in the comment as well

ChengYanJin replied:
> Something wrong with this? Ticket in `core-ui` maybe? If there's one already, having it in the comment can help :)

Oh. Sorry, I should have removed those tooltip things. 
At first, I wanted to add a tooltip for the trash button. 
We may need the tooltip later, but I created a ticket in core-ui already!

gdemonet commented on ui/src/containers/SolutionList.js:
```diff
@@ -135,56 +165,68 @@ const SolutionsList = props => {
+            return (
+              <span
+                key={idx}
+              >{`${deployedSolution.name}(v.${deployedSolution.version})`}</span>
```
minor: let's show a space after the solution name
```suggestion
              >{`${deployedSolution.name} (v.${deployedSolution.version})`}</span>
```

ChengYanJin committed:
ui/solutions: Add the validation schema of the environment name

- Most k8s resource types require a name that can be used as a DNS subdomain name as defined in RFC 1123. This means the name must:
  - contain no more than 253 characters
  - contain only lowercase alphanumeric characters, ‘-’ or ‘.’
  - start with an alphanumeric character
  - end with an alphanumeric character
- Reuse the volume name schema as they share the same rule. Rename of the key `volume_name_error` to `name_error`
  as the error message is the same.

Refs: #2278

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -67,113 +80,257 @@ export function createEnvironmentAction(newEnvironment) {
+    ?.filter(job => job.name.startsWith('prepare-env/') && !job.completed)
+    .map(job => job.name.replace(/^(prepare-env\/)/, ''));
+  const environments = yield call(SolutionsApi.listEnvironments);
+  const updatedEnvironments = yield call(updateEnvironments, environments);
```
Is it useful to split this "hydration" step from the original `listEnvironments`, in the end?

ChengYanJin replied:
In the end, I still prefer to sperate the `updateEnvironments` from `fetchEnvironments`.
For me the purpose of updateEnvironments is to update the Environment with the deployed solutions it's a bit more than fetchEnvironments IMO

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -67,113 +80,257 @@ export function createEnvironmentAction(newEnvironment) {
+    const resultCreateNamespacedConfigMap = yield call(
+      SolutionsApi.createNamespacedConfigMap,
+      name,
+    );
```
Could be part of `SolutionsApi.createEnvironment` I believe. Especially since we're not sending any specific information to the user about an error, would it occur.

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -67,113 +80,257 @@ export function createEnvironmentAction(newEnvironment) {
+export function* prepareEnvironment(action) {
+  const { envName, solution } = action.payload;
+  const addedSolutionName = solution.solution.value;
+  const addedSolutionVersion = solution.version.value;
```
We're extracting `.value` because it comes from a form field? I'd argue that's not the sagas' responsibility.
In general, we should strive to only give each function what it needs and nothing else. In this case, extracting the `.value`s into a usable payload should be done in the component which displays the form and triggers the saga.

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -67,113 +80,257 @@ export function createEnvironmentAction(newEnvironment) {
+      addNotificationErrorAction({
+        title: intl.translate('prepare_environment'),
+        message: intl.translate('env_preparation_failed', { envName }),
+      }),
```
This was an error when adding the Solution to the Env's ConfigMap. Maybe another message could be useful here.

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -67,113 +80,257 @@ export function createEnvironmentAction(newEnvironment) {
+      const solutions = Object.keys(envConfig);
+      // we may have several soutions in one environment
+      for (const solution of solutions) {
+        const solutionOperatorDeployment = yield call(
+          CoreApi.getNamespacedDeployment,
+          `${solution}${OPERATOR_}`,
+          env.name,
+        );
+        const solutionUIDeployment = yield call(
+          CoreApi.getNamespacedDeployment,
+          `${solution}${UI_}`,
+          env.name,
+        );
```
Mmmh, we should not rely on the name of these resources, as it may change. Instead, let's use sufficient selectors (cc @alexandre-allard-scality - maybe you know which ones we set?).

alexandre-allard replied:
I don't see any useful label for this, especially for the Operator which contains only a single label refering to the MetalK8s cluster version:
```
metadata:
  [...]
  labels:
    metalk8s.scality.com/version: 2.4.3-dev
```
For the UI part, there are labels, but they're all computed on the same basis as the deployment name... so, not more reliable than the deployment name:
```
metadata:
  [...]
  labels:
    app: example-solution-ui
    app.kubernetes.io/component: ui
    app.kubernetes.io/managed-by: salt
    app.kubernetes.io/name: example-solution-ui
    app.kubernetes.io/part-of: example-solution
    app.kubernetes.io/version: 0.1.0-dev
    heritage: example-solution
    metalk8s.scality.com/version: 2.4.3-dev
```
Computed from:
```
metadata:
  name: {{ solution }}-ui
  namespace: {{ namespace }}
  labels:
    app: {{ solution }}-ui
    heritage: {{ solution }}
    app.kubernetes.io/name: {{ solution }}-ui
    app.kubernetes.io/version: {{ version }}
    app.kubernetes.io/component: ui
    app.kubernetes.io/part-of: {{ solution }}
    # UIs are deployed and managed by Salt, provided with MetalK8s
    app.kubernetes.io/managed-by: salt
```
Should we had some labels to these ?

ChengYanJin replied:
Response from Alexandre:
I don't see any useful label for this, especially for the Operator which contains only a single label refering to the MetalK8s cluster version...

For the UI part, there are labels, but they're all computed on the same basis as the deployment name... so, not more reliable than the deployment name:
```
  labels:
    app: example-solution-ui
    app.kubernetes.io/component: ui
    app.kubernetes.io/managed-by: salt
    app.kubernetes.io/name: example-solution-ui
    app.kubernetes.io/part-of: example-solution
    app.kubernetes.io/version: 0.1.0-dev
    heritage: example-solution
    metalk8s.scality.com/version: 2.4.3-dev
```
Computed from:

```
metadata:
  name: {{ solution }}-ui
  namespace: {{ namespace }}
  labels:
    app: {{ solution }}-ui
    heritage: {{ solution }}
    app.kubernetes.io/name: {{ solution }}-ui
    app.kubernetes.io/version: {{ version }}
    app.kubernetes.io/component: ui
    app.kubernetes.io/part-of: {{ solution }}
    # UIs are deployed and managed by Salt, provided with MetalK8s
    app.kubernetes.io/managed-by: salt
```

gdemonet replied:
These labels are what we should rely on IMO, not the names (we may want to change them for whatever reason). If I'm looking for _the UI of Solution "toto" in version "1.0.0", deployed in the Env namespace "prod"_, I'll do the query:

```shell
kubectl get deploy -n prod \
    -l app.kubernetes.io/part-of=toto \
    -l app.kubernetes.io/component=ui \
    -l app.kubernetes.io/version=1.0.0
```

ChengYanJin replied:
I will create a debt ticket in order to check if we have the deployed UI and Solution service

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -67,113 +80,257 @@ export function createEnvironmentAction(newEnvironment) {
-  yield put(setSolutionsRefeshingAction(false));
+export function* deleteEnvironment(action) {
+  const envName = action.payload;
+  const result = yield call(CoreApi.deleteEnvironment, envName);
```
This `deleteEnvironment` should live in `SolutionsApi`, not `CoreApi`. "Environments" are not even a concept in K8s world, it's just our creation :)

ChengYanJin committed:
ui/test: Add unit test for delete Environment

Refs: #2278

gdemonet commented on ui/src/services/k8s/core.js:
```diff
@@ -0,0 +1,129 @@
+    );
+  } catch (error) {
+    return { error };
+  }
```
I think there was a misunderstanding here. JSON Patch **is** supported by K8s API. You just need to send it with the right header:

```javascript
export async function patchNamespacedConfigMap(name, namespace, { jsonPatch, mergePatch }) {
  if (jsonPatch !== undefined) {
    cTypeHeader = 'application/json-patch+json';
    body = jsonPatch;
  } else {
    cTypeHeader = 'application/merge-patch+json';
    body = mergePatch;
  }

  try {
    return await coreV1.patchNamespacedConfigMap(
      name,
      namespace,
      body,
      undefined,
      undefined,
      { headers: { 'Content-Type': cTypeHeader } },
    );
  } catch (error) {
    return { error };
  }
}
```

gdemonet replied:
(which means you could also get rid of the `jsonpatch` dependency)

gdemonet commented on ui/src/services/k8s/core.js:
```diff
@@ -0,0 +1,129 @@
+  }
+}
+
+export async function getNamespacedDeployment(name, namespace) {
+  try {
+    return await appsV1.readNamespacedDeployment(name, namespace);
+  } catch (error) {
+    return { error };
+  }
+}
```
Not convinced about the need for such transparent helpers, but we can leave it as is.

gdemonet reviewed (changes requested):
Some general comments on usability of the feature (most of these remarks can be dispatched in future iterations):

- `solutions.sh import -a /path/to/my.iso` doesn't put it in the `metalk8s-solutions` ConfigMap, which means UI needs a call to `solutions.sh activate ...` before seeing a newly added Solution ISO (that's not related to UI, more of a Salt problem I think)
- upgrading a Solution instance in an Environment is tricky: you need to **Add** the new version, which will override the old one - IMO, that's counter-intuitive
- there's a bug in the "add Solution to Env" modal: if I have Solution 1 with version A, switching to Solution 2 will keep version A selected, even if there's no version A for Solution 2 - I can submit the form, and the Salt job then fails because the version is not available
- notifications are bugged for me: when a "prepare" job fails, I get the error notification twice - when it succeeds, I get  **four** success notifications at once!
- not sure why exactly, but I deployed two Solution instances in my environment, and only one shows up in the table (the first one I deployed)

ChengYanJin committed:
ui/solutions: Fix the bug for Formik form

- once we select a solution, we should update the initialValues of version

Refs: #2278

ChengYanJin committed:
ui/solutions: Fix the display bug when there are more than one solutions deployed

- When there is solution in the solutions array, we don't need to initialize the solutions[]

Refs: #2278

ChengYanJin committed:
ui/solutions: Fix the bug that it displays multiple notifications at the same time

- Remove the watchJob function use notifyDeployJobCompleted to show the notification instead of
  The reason display multiple notification inside the watchjob function,so that we got all the job list status from Salt
  However, Salt Garbage collection of jobs for Redux state in every 5mins. So we display
  all the job in these 5 mins.

- Remove the name for job use job.type instead of

- For the error notication we display the error happen in which step and reason instead of only an error message

Refs: #2278

ChengYanJin committed:
ui/solutions: Specific the error when preparing the Environment has failed is because of adding solution to configMap

- extracting values of solName and solVersion before sending them to saga

Refs: #2278

ChengYanJin committed:
ui/solutions: Refactoring the createEnvironment saga

ChengYanJin commented:
> Some general comments on usability of the feature (most of these remarks can be dispatched in future iterations):
> 
> * `solutions.sh import -a /path/to/my.iso` doesn't put it in the `metalk8s-solutions` ConfigMap, which means UI needs a call to `solutions.sh activate ...` before seeing a newly added Solution ISO (that's not related to UI, more of a Salt problem I think)
> * upgrading a Solution instance in an Environment is tricky: you need to **Add** the new version, which will override the old one - IMO, that's counter-intuitive
> * there's a bug in the "add Solution to Env" modal: if I have Solution 1 with version A, switching to Solution 2 will keep version A selected, even if there's no version A for Solution 2 - I can submit the form, and the Salt job then fails because the version is not available
> * notifications are bugged for me: when a "prepare" job fails, I get the error notification twice - when it succeeds, I get  **four** success notifications at once!
> * not sure why exactly, but I deployed two Solution instances in my environment, and only one shows up in the table (the first one I deployed)

Thanks for the review!
Please find the following responses which map to each of the points.

-  Alexandre will help me to handle this, will do in a sperate PR
-  I got your point, how do we handle the upgrading the solution instances I will create a ticket and do it in another PR.
- Fix the bug in `ae07f9b2 - ui/solutions: Fix the bug for Formik form`
- Fix the bug in `ff7fc846 - ui/solutions: Fix the bug that it displays multiple notifications at the same time`
- Fix the bug in `045f3feb - ui/solutions: Fix the display bug when there are more than one solutions deployed`

For the other's comments, I addressed them all but the jsonPatch, for some reason, I got 422 when I patch...

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -330,12 +297,41 @@ export function* deleteEnvironment(action) {
     takeEvery(CREATE_ENVIRONMENT, createEnvironment),
     takeEvery(PREPARE_ENVIRONMENT, prepareEnvironment),
     takeEvery(DELETE_ENVIRONMENT, deleteEnvironment),
+    takeEvery(JOB_COMPLETED, notifyDeployJobCompleted),
```
Obviously! Don't know why I made my own with `while (true) { take(JOB_COMPLETED); ... }`... :man_facepalming:

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -175,8 +177,10 @@ export function* prepareEnvironment(action) {
       addNotificationErrorAction({
-        title: intl.translate('prepare_environment'),
-        message: intl.translate('env_preparation_failed', { envName }),
+        title: intl.translate('add_solution_to_configMap', {
```
Let's be consistent and keep translation keys always lowercase please :) so `configmap`, not `configMap` (here and everywhere else, of course)

gdemonet commented:
> the jsonPatch, for some reason, I got 422 when I patch...

What's the error message? Maybe the patch body is not in the right format, or some keys can't be patched as we'd like to...

ChengYanJin committed:
ui/solutions: Apply json patch provided by k8s api to patch solution for configmap

Although we coulf create empty data field when we create confimap, it may be removed by k8s.
So the for the first patch we need to create the data object,
for the next patches we can use the path `/data/${solName}`.

Refs: #2278

ChengYanJin commented:
> > the jsonPatch, for some reason, I got 422 when I patch...
> 
> What's the error message? Maybe the patch body is not in the right format, or some keys can't be patched as we'd like to...

Done with`12ca05f3 - ui/solutions: Apply json patch provided by k8s api to patch solution for configmap`

ChengYanJin commented:
> > > the jsonPatch, for some reason, I got 422 when I patch...
> > 
> > 
> > What's the error message? Maybe the patch body is not in the right format, or some keys can't be patched as we'd like to...
> 
> Done with `12ca05f3 - ui/solutions: Apply json patch provided by k8s api to patch solution for configmap`

slaperche-scality commented on ui/src/services/k8s/core.js:
```diff
@@ -79,10 +78,12 @@ export async function queryPodInNamespace(namespace, podLabel) {
 
 export async function createNamespacedConfigMap(name, namespace, restProps) {
   const body = {
+    // it actually be removed by the k8s when the data is empty.
```
Since it's removed, maybe you don't need to put it (anyway you have to create it with the first patch)

slaperche-scality commented on ui/src/services/k8s/solutions.js:
```diff
@@ -119,7 +119,17 @@ export async function getEnvironmentConfigMap(environment) {
+    namespace,
+  );
+  let patch;
+  // the first patch we need to create data key. otherwise we will get 422 patch error.
```
Or simply: `// Create the `data` key if it doesn't exists, otherwise we can't insert the solution field inside`

slaperche-scality commented on ui/src/services/k8s/core.js:
```diff
@@ -79,10 +78,12 @@ export async function queryPodInNamespace(namespace, podLabel) {
 
 export async function createNamespacedConfigMap(name, namespace, restProps) {
```
In the commit message:

- `create the data object,` => `create the field "data"` (object is a bit ambiguous in this context).
- `we could using the path` => `we can use the path`

slaperche-scality reviewed (commented):
Some minor comments (mainly about comments ^^)

ChengYanJin commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

gdemonet reviewed (approved):
Didn't test manually again, but let's merge it and test all of this automatically in post-merge :)

Great job @ChengYanJin ! :clap:

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/1077b650533a3a0139e9b1deb848e657e1ba3769) did not succeed in branch feature/2278-update-solutions-ui.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye chengyanjin.
